gcloud-java-storage/src/main/java/com/google/gcloud/storage/ObjectId.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2015 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.google.gcloud.storage;
+
+import java.net.URI;
+import java.util.Objects;
+
+/**
+ * Identifier for a stored object.
+ */
+public class ObjectId {
+
+    private final String bucketName;
+    private final String objectName;
+    private final URI uri;
+
+    public ObjectId(String uri) {
+        this(URI.create(uri));
+    }
+
+    /**
+     * Constructs an ObjectId from a {@code gsutil} style URI.
+     *
+     * @param uri the object's URI
+     */
+    public ObjectId(URI uri) {
+        Objects.requireNonNull(uri, "uri must not be null");
+        this.uri = uri;
+        if (uri.isAbsolute()) {
+            if (!"gs".equals(uri.getScheme())) {
+                throw new IllegalArgumentException("uri must use gs: scheme");
```

Comment:
I think this should also be part of the discussion for the intent of considering/supporting URI in our API.
For example if it is going to be more than just a another way for a user to keep a reference to bucket+blobName than
maybe the spi layer should decide how to parse it (and e.g. verify its scheme,...).